server/routes/shared_api/node.py

```diff
@@ -45,3 +51,59 @@ def get_property_value(direction):
     prop = request.json['prop']
   response = fetch.raw_property_values(dcids, prop, direction == 'out')
   return Response(json.dumps(response), 200, mimetype='application/json')
+
+
+@bp.route('/propvals', methods=['GET', 'POST'])
+def expression_property_value():
+  """
+  Returns the property values for given node dcids and a relation expression for
+  a property (https://docs.datacommons.org/api/rest/v2#relation-expressions)
+  """
+  dcids = request.args.getlist('dcids')
+  if not dcids:
+    dcids = request.json['dcids']
+  prop_expression = request.args.get('propExpr')
+  if not prop_expression:
+    prop_expression = request.json['propExpr']
+  if not dcids:
+    return 'error: must provide a `dcids` field', 400
+  if not prop_expression:
+    return 'error: must provide a `propExpr` field', 400
+  response = get_expression_property_value(dcids, prop_expression)
+  return Response(json.dumps(response), 200, mimetype='application/json')
+
+
+def get_expression_property_value(dcids, expression):
+  """
+  Returns the property values for given node dcids and a relation expression for
+  a property (https://docs.datacommons.org/api/rest/v2#relation-expressions)
+  """
+  # Get a list of single property relation expressions (the expression passed to
+  # this function can chain multiple relation expressions)
+  expression_matches = re.finditer(_PROPERTY_EXPRESSION_RE, expression)
+  results = {}
+  # the list of dcids to use to fetch the next round of property values
+  curr_dcids = dcids
+  # map of dcids from curr_dcids to the original dcid passed to this function
+  curr2orig_dcids = {dcid: dcid for dcid in curr_dcids}
+  for match in expression_matches:
```

Comment:
I am a bit confused of the logic here. Can you comment with a concrete example and the steps to do the data fetch?

Reply:
not sure if you wanted an example here or in the code, but I'll give the example here and I can add it to the code as well:

node dcids: [rs13317, rs7903146]
prop expression: <-referenceSNPClusterID{typeOf:GeneticVariantGeneAssociation}->geneSymbol

1. we have 2 expression matches:
     - <-referenceSNPClusterID{typeOf:GeneticVariantGeneAssociation}
     - ->geneSymbol
2. For the first expression match, we fetch data using:
`fetch.raw_property_values(nodes=[rs13317, rs7903146], prop=referenceSNPClusterID, out=False, constraints='{typeOf:GeneticVariantGeneAssociation}')`
3. Say from that data fetch, we get back:
`{rs13317: [{dcid: a}, {dcid: b}], rs7903146: [{dcid: e}, {dcid: f}]}`
6. For the next expression match, we fetch data using 
`fetch.raw_property_values(nodes=[a, b, e, f], prop=geneSymbol, out=True, constraints='')`
7. Say from that data fetch, we get back 
`{a: [{dcid: aa}, {dcid: bb}], b: [{dcid: cc}, {dcid: dd}], e: [{dcid: ee}, {dcid: ff}], f: [{dcid: gg}, {dcid: hh}]}`
8. In the final result, we want
`{rs13317:  [{dcid: aa}, {dcid: bb}, {dcid: cc}, {dcid: dd}], rs7903146: [{dcid: ee}, {dcid: ff}, {dcid: gg}, {dcid: hh}]}`

Reply:
I see. thanks for the detailed example! I wonder if you want to have 

- one function that: parse expression, fetch data and return dcid2dcid map.
- one function that takes a list of such mappings and join them together to get the final result?

Reply:
I tried splitting out the functions but not sure if it's how you envisioned